app/src/main/java/app/revanced/integrations/patches/components/LithoFilterPatch.java

```diff
@@ -0,0 +1,263 @@
+
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.Spliterator;
+import java.util.function.Consumer;
+
+import app.revanced.integrations.settings.SettingsEnum;
+import app.revanced.integrations.utils.LogHelper;
+import app.revanced.integrations.utils.ReVancedUtils;
+
+abstract class FilterGroup<T> {
+    final static class FilterGroupResult {
+        private final boolean filtered;
+        private final SettingsEnum setting;
+
+        public FilterGroupResult(final SettingsEnum setting, final boolean filtered) {
+            this.setting = setting;
+            this.filtered = filtered;
+        }
+
+        public SettingsEnum getSetting() {
+            return setting;
+        }
+
+        public boolean isFiltered() {
+            return filtered;
+        }
+    }
+
+    protected final SettingsEnum setting;
+    protected final T[] filters;
+
+    /**
+     * Initialize a new filter group.
+     *
+     * @param setting The associated setting.
+     * @param filters The filters.
+     */
+    @SafeVarargs
+    public FilterGroup(final SettingsEnum setting, final T... filters) {
+        this.setting = setting;
+        this.filters = filters;
+    }
+
+    public boolean isEnabled() {
+        return setting.getBoolean();
+    }
+
+    public abstract FilterGroupResult check(final T stack);
+}
+
+class StringFilterGroup extends FilterGroup<String> {
+
+    /**
+     * {@link FilterGroup#FilterGroup(SettingsEnum, Object[])}
+     */
+    public StringFilterGroup(final SettingsEnum setting, final String... filters) {
+        super(setting, filters);
+    }
+
+    @Override
+    public FilterGroupResult check(final String string) {
+        return new FilterGroupResult(setting, string != null && ReVancedUtils.containsAny(string, filters));
+    }
+}
+
+final class CustomFilterGroup extends StringFilterGroup {
+
+    /**
+     * {@link FilterGroup#FilterGroup(SettingsEnum, Object[])}
+     */
+    public CustomFilterGroup(final SettingsEnum setting, final SettingsEnum filter) {
+        super(setting, filter.getString().split(","));
+    }
+}
+
+class ByteArrayFilterGroup extends FilterGroup<byte[]> {
+    // Modified implementation from https://stackoverflow.com/a/1507813
+    private int indexOf(final byte[] data, final byte[] pattern) {
+        // Computes the failure function using a boot-strapping process,
+        // where the pattern is matched against itself.
+
+        final int[] failure = new int[pattern.length];
+
+        int j = 0;
+        for (int i = 1; i < pattern.length; i++) {
+            while (j > 0 && pattern[j] != pattern[i]) {
+                j = failure[j - 1];
+            }
+            if (pattern[j] == pattern[i]) {
+                j++;
+            }
+            failure[i] = j;
+        }
+
+        // Finds the first occurrence of the pattern in the byte array using
+        // KNP matching algorithm.
```

Comment:
Isn't it `KMP` instead of `KNP` algorithm?

Reply:
Yes: https://de.wikipedia.org/wiki/Knuth-Morris-Pratt-Algorithmus